qemu/tests/cpu_device_hotplug.py

```diff
@@ -1,100 +1,119 @@
-        output = vm.monitor.send_args_cmd("info cpus")
-        logging.debug("Output of info CPUs:\n%s", output)
-
-        cpu_regexp = re.compile(r"CPU #(\d+)")
-        total_cpus_monitor = len(cpu_regexp.findall(output))
-        if total_cpus_monitor != total_cpus:
-            test.fail("Monitor reports %s CPUs, when VM should have"
-                      " %s" % (total_cpus_monitor, total_cpus))
-        error_context.context("hotplugging finished, let's wait a few sec and"
-                              " check CPUs quantity in guest.", logging.info)
-        if not utils_misc.wait_for(lambda: cpu.check_if_vm_vcpu_match(
-                                   total_cpus, vm),
-                                   60 + total_cpus, first=10,
-                                   step=5.0, text="retry later"):
-            test.fail("CPU quantity mismatch cmd after hotplug !")
-        error_context.context("rebooting the vm and check CPU quantity !",
+    def get_vm_cpu_ids():
+        cmd = "grep processor /proc/cpuinfo"
+        output = session.cmd_output(cmd)
+        return set(re.findall(r"\d+", output))
+
+    def sub_reboot():
+        error_context.context("Reboot guest after vcpu %s" % plug_test,
                               logging.info)
-        session = vm.reboot()
-        if not cpu.check_if_vm_vcpu_match(total_cpus, vm):
-            test.fail("CPU quantity mismatch cmd after hotplug and reboot !")
+        vm.reboot(method=params["reboot_method"])
 
-    error_context.context("boot the vm, with '-smp X,maxcpus=Y' option,"
-                          "thus allow hotplug vcpu", logging.info)
+    def sub_shutdown():
+        error_context.context("Shutdown guest after vcpu %s" % plug_test,
+                              logging.info)
+        shutdown_method = params["shutdown_method"]
+        if shutdown_method == "shell":
+            session.sendline(params["shutdown_command"])
+            error_context.context("waiting VM to go down (guest shell cmd)",
+                                  logging.info)
+        elif shutdown_method == "system_powerdown":
+            vm.monitor.system_powerdown()
+            error_context.context("waiting VM to go down (qemu monitor cmd)",
+                                  logging.info)
+        if not vm.wait_for_shutdown(360):
+            test.fail("Guest refuses to go down.")
+
+    def sub_migrate():
+        error_context.context("Migrate guest after vcpu %s" % plug_test,
+                              logging.info)
+        vm.migrate(float(params.get("mig_timeout", "3600")))
+
+    def sub_online_offline():
+        error_context.context("Online/Offline guest CPUs after vcpu %s"
+                              % plug_test, logging.info)
+        cpu_ids = list(current_guest_cpu_ids - guest_cpu_ids)
+        cpu_ids.sort()
+        cmd = "echo %d > /sys/devices/system/cpu/cpu%s/online"
+        for cpu_id in cpu_ids:
+            try:
+                session.cmd(cmd % (0, cpu_id))
```

Comment:
Should we check if cpu is offlined successfully?

Reply:
updated